File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        // pair -> height, idx
        vector<pair<int,int>> h(n);
        for(int i=0;i<n;i++){
            h.push_back({heights[i],i});
        }
        
        stack<pair<int,int>> s;
        
        vector<int> prevSmaller(n);
        
        for(int i=0;i<n;i++){
            auto cur = heights[i];
            while(!s.empty()&&s.top().first>=cur){
                s.pop();
            }
            if(!s.empty()){
                prevSmaller[i] = s.top().second;
            }
            else{
                prevSmaller[i] = -1;
            }
            s.push({cur,i});
        }
        
        while(!s.empty()){
            s.pop();
        }
        
        int maxArea = 0;
        for(int i=n-1;i>=0;i--){
            auto cur = heights[i];
            int temp;
            while(!s.empty()&&s.top().first>=cur){
                s.pop();
            }
            if(!s.empty()){
                temp = s.top().second;
            }
            else{
                temp = n;
            }
            
            auto tempans = cur*(temp-prevSmaller[i]-1);
            if(tempans>maxArea){
                maxArea = tempans;
            }
            s.push({cur,i});
        }
        
        return maxArea;
        
    }
};

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        
        st = deque()
        # def getNextSmaller(arr,init):
        nextSmaller = [n]*n
        for i in range(n-1,-1,-1):
            while st and heights[st[-1]]>=heights[i]:
                st.pop()
            if st:
                nextSmaller[i]=st[-1]
            st.append(i)
        # return nextSmaller
        prevSmaller = [-1 for i in range(n)]
        st = []
        for i in range(n):
            while st and heights[st[-1]]>=heights[i]:
                st.pop()
            if st:
                prevSmaller[i] = st[-1]
            st.append(i)
        ans = 0
        # nextSmaller = getNextSmaller(heights,n)
        # print(prevSmaller)
        # print(nextSmaller)
        for idx in range(n):
            cur = heights[idx]
            t = nextSmaller[idx]
            s = prevSmaller[idx]
            L,R = max(0,idx-s)*cur,max(0,t-idx)*cur
            ans = max(ans,L+R-cur)
        return ans